lib/firmware.ex

```diff
@@ -121,6 +128,13 @@ defmodule Nerves.Firmware do
   """
   @spec upgrade_and_finalize(String.t, [binary]) :: :ok | {:error, reason}
   def upgrade_and_finalize(firmware, args \\ []) do
+    pub_key_path = Application.get_env(:nerves_firmware, :pub_key_path)
+    args = if pub_key_path do
+      IO.puts "using signature!"
```

Comment:
Same comment as line 97 regarding IO.puts.

Since this whole code block is duplicated both here an in apply(), consider moving both blocks to a shared private function, so the final GenServer.call becomes something like..
```elixir
GenServer.call @server, {:upgrade_and_finalize, firmware, args ++ pub_key_args()}, :infinity
```